Add workflow to build nightly releases

ddoktorski opened the pull request:
<!-- Reference any GitHub issues resolved by this PR -->

Closes #3300

ddoktorski committed:
Add workflow to build nightly releases

ddoktorski commented:
Example run: https://github.com/ddoktorski/starknet-foundry/actions/runs/15210212174
Example release: https://github.com/ddoktorski/starknet-foundry-nightlies/releases/tag/nightly-2025-05-20-15210212174 (run id included in the name only for testing)

ddoktorski commented:
TODO:
* Set `SNFOUNDRY_NIGHTLIES_CONTENTS_WRITE` secret in the repository as a token with write permission to https://github.com/foundry-rs/starknet-foundry-nightlies

kkawula commented:
#3423

kkawula commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,187 @@
+on:
+  workflow_dispatch:
+    inputs:
+      dry_run:
+        description: "Dry run"
+        type: boolean
+        default: true
```
As it has default set to true, I don't think so

cptartur commented on .github/workflows/_build-binaries.yml:
```diff
@@ -7,6 +7,9 @@ on:
       version:
         required: true
         type: string
+      ref:
```
Don't we want to use some more descriptive name?

ddoktorski replied:
It seems natural to name it `ref`, as the same name is used by [actions/checkout](https://github.com/actions/checkout?tab=readme-ov-file#checkout-a-different-branch)

kkawula commented on .github/workflows/_publish-plugin.yml:
```diff
@@ -28,10 +28,10 @@ jobs:
           
           snforge_scarb_plugin_version=$(get_version "${{ inputs.overridden_plugin_version }}")
           
+          # TODO: It should check dev registry if prod_registry is false
```
https://github.com/foundry-rs/starknet-foundry/pull/3423

kkawula committed:
Add Ma'at trigger step to nightly workflow (#3423)

<!-- Reference any GitHub issues resolved by this PR -->

Closes #

## Introduced changes

<!-- A brief description of the changes -->

-

## Checklist

<!-- Make sure all of these are complete -->

- [ ] Linked relevant issue
- [ ] Updated relevant documentation
- [ ] Added relevant tests
- [ ] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

---------

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

ddoktorski commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,187 @@
+
+      - name: Create source code archives
+        run: |
+          git archive "--prefix=starknet-foundry-${{ needs.prepare.outputs.nightly_tag }}/" -o "starknet-foundry-${{ needs.prepare.outputs.nightly_tag }}.zip" HEAD
```
Actually, it seems like more effort to add specific logic to exclude Windows than to just keep it. I would opt to leave it for now and we can always adjust it later if there are any maintenance issues.

cptartur replied:
We can remove windows entirely from build workflows. @kkawula will do so as a prerequisite to #3440 

But to avoid conflicts, you can keep it for now.

ddoktorski commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,187 @@
+            # starknet-foundry-nightly-2025-05-22-aarch64-apple-darwin.tar.gz
+
+            if ! [[ $(grep "#" <<< $file) ]]; then
+              label=$(echo "$file" | sed -E "s/v[^+]*\+//" | sed -E "s|.*/||")
```
Yes, they are allowed in extended regular expressions: https://www.gnu.org/software/sed/manual/sed.html#ERE-syntax

cptartur replied:
TIL that you can use any value as delimiter in sed https://backreference.org/2010/02/20/using-different-delimiters-in-sed/index.html
Doesn't seem to require `-E` mode, I think it can be removed in case of your regex.

ddoktorski replied:
I didn't go into much detail but doing for example
```
echo "starknet-foundry-v0.44.0+nightly-2025-05-22-aarch64-apple-darwin.tar.gz" | sed "s/v[^+]*\+//"
```
without `-E` fails

cptartur replied:
Okay, doesn't matter that much, we can keep it

ddoktorski committed:
Remove job to run maat experiment

ddoktorski committed:
Merge branch 'master' into 3300-nightly-release-workflow

franciszekjob commented on RELEASING.md:
```diff
@@ -13,3 +13,16 @@ with the name format `vMAJOR.MINOR.PATCH`.
+Nightly release must be triggered manually using the Nightly GitHub action.
+This action builds binaries from specified ref and uploads them to the [starknet-foundry-nightlies](https://github.com/foundry-rs/starknet-foundry-nightlies) repository.
+Additionally, there are `stds` and `plugin` version uploaded to the [_dev_ registry](https://scarbs.dev/).
+After a successful release, [Ma'at](https://github.com/software-mansion/maat) is automatically trigger to run experiments in nightly workspace. Results can be found [here](https://docs.swmansion.com/maat/)
```
```suggestion
After a successful release, [Ma'at](https://github.com/software-mansion/maat) is automatically triggered to run experiments in nightly workspace. Results can be found [here](https://docs.swmansion.com/maat/)
```

franciszekjob commented on RELEASING.md:
```diff
@@ -13,3 +13,16 @@ with the name format `vMAJOR.MINOR.PATCH`.
+
+Nightly release must be triggered manually using the Nightly GitHub action.
+This action builds binaries from specified ref and uploads them to the [starknet-foundry-nightlies](https://github.com/foundry-rs/starknet-foundry-nightlies) repository.
+Additionally, there are `stds` and `plugin` version uploaded to the [_dev_ registry](https://scarbs.dev/).
```
```suggestion
Additionally, there are `stds` and `plugin` versions uploaded to the [_dev_ registry](https://scarbs.dev/).
```

cptartur commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,186 @@
+            # becomes:
+            # starknet-foundry-nightly-2025-05-22-aarch64-apple-darwin.tar.gz
+
+            if ! [[ $(grep "#" <<< $file) ]]; then
```
Why do we only change the files that don't include `#`?

ddoktorski replied:
We rename build artifacts to remove the version from their names. Files that include `#` are [source code files](https://github.com/foundry-rs/starknet-foundry/pull/3400/files/5df1d87dc762b0cf8d4df8bf047f66c2e8dc4352#diff-0d5658b415099a82c11c03a06ca4ec765b4003a1f4b2f3f1943980a882cf8aa6R153-R154) and do not need to be renamed.

cptartur replied:
If there are only two source files that need to be handled separately, maybe it'd be more readable if we had two different loops rather than adding this if?

cptartur replied:
It'd be a slight duplication of code but I think it would be clearer to understand

ddoktorski replied:
Done

ddoktorski committed:
Apply suggestions from code review

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

ddoktorski committed:
Merge branch 'master' into 3300-nightly-release-workflow

ddoktorski committed:
Split release assets upload into 2 steps
